src/cirrus/deploy_plugins.py

```diff
@@ -0,0 +1,195 @@
+    """
+    _get_plugin_
+
+    Helper to get the plugin instance based on the name
+    """
+    return _PLUGINS.get(plugin_name)
+
+
+class DeployerRegistry(type):
+    """
+    _DeployerRegistry_
+
+    metaclass to register plugin subclasses with plugins registry
+
+    """
+    def __init__(cls, name, bases, dct):
+        pname = getattr(cls, 'plugin_name')
+        if pname is not None:
+            if pname not in _PLUGINS:
+                _PLUGINS[pname] = cls()
+            else:
+                raise RuntimeError(
+                    "Duplicate plugin name: {0}".format(pname)
+                )
+        super(DeployerRegistry, cls).__init__(name, bases, dct)
+
+
+class DeployerPlugin(object):
+    """
+    _DeployerPlugin_
+
+    Base class for Deployer plugins.
+
+    Plugin implementations should inherit this,
+    override the plugin_name attribute (this will be exposed
+        as the command name in the CLI, for example plugin_name="womp"
+        will make the plugin available via git cirrus deploy womp )
+
+    The build_parser method sets up the CLI argparse options for the plugin
+    The deploy method consumes the options namespace from the parser
+    and then executes whatever code it needs to perform the deployment action
+
+    Registry with the plugin framework is handled via the metaclass
+
+    The base class provides some helpers to get at the package
+    and user configuration to allow for config access and customisation
+
+
+    """
+    __metaclass__ = DeployerRegistry
+    plugin_name = None
+
+    def __init__(self):
+        self.package_conf = load_configuration()
+
+    def deploy(self, options):
+        """
+        _deploy_
+
+        :param options: Instance of argparse namespace containing
+        the command line options values
+
+        """
+        raise NotImplementedError(
+            "{0}.deploy not implemented".format(type(self).__name__)
+        )
+
+    def build_parser(self, command_parser):
+        """
+        _build_parser_
+
+        Hook for the plugin to build out its commandline tool
+        suite.
+
+        command_parser is the argparse parser instance for the plugin
+        to add its extensions to
+        """
+        raise NotImplementedError(
+            "{0}.build_parser not implemented".format(type(self).__name__)
+        )
+
+
+class ChefServerDeployer(DeployerPlugin):
+    """
+    _ChefServerDeployer_
+
+    Implement a chef deployment plugin.
+
+    This assumes your deployment process looks something like:
+
+    1. Bump the package version in an env/role/node for some attribute
+    2. Run pre-chef-client commands on the specified nodes
+    3. Run chef-client on the specified nodes
+    4. Run post-chef-client commands on the specified nodes
+
+    This plugin runs chef-client serially and is more suited to
+    a few nodes (eg a CI or integ system) rather than a full
+    scale deployment across many nodes
+
+    """
+    plugin_name = 'chef'
```

Comment:
I like this pattern because I don't have to remember to manually register my plugin with some dict elsewhere in the file (or in another file). But if you forget to set it, you have a fairly subtle bug (the plugin just fails to register, and you don't know why it isn't there).

I'd be tempted to try to do something like this in the metaclass:

    if cls.plugin_name is None:
        cls.plugin_name = name[:-5] if name.endswith("Plugin").lower() else name.lower()

Then you'd have to skip adding "Deployer" to the registry. But maybe that is too much magic.